src/Parser/PotreeBinParser.js

```diff
@@ -1,61 +1,118 @@
+    NORMAL_SPHEREMAPPED: {
+        numElements: 2,
+        arrayType: Uint8Array,
+        attributeName: 'sphereMappedNormal',
+    },
+    // see https://web.archive.org/web/20150303053317/http://lgdv.cs.fau.de/get/1602
+    NORMAL_OCT16: {
+        numElements: 2,
+        arrayType: Uint8Array,
+        attributeName: 'oct16Normal',
+    },
+    NORMAL: {
+        numElements: 3,
+        arrayType: Float32Array,
+        attributeName: 'normal',
+    },
+};
+
 // Parse .bin PotreeConverter format
 export default {
     /** @module PotreeBinParser */
     /** Parse .bin PotreeConverter format and convert to THREE.Points
      * @function parse
      * @param {ArrayBuffer} buffer - the bin buffer.
+     * @param {Object} pointAttributes - the point attributes information contained in layer.metadata coming from cloud.js
      * @return {Promise} a promise that resolves with a THREE.Points.
      *
      */
-    parse: function parse(buffer) {
+    parse: function parse(buffer, pointAttributes) {
         if (!buffer) {
             throw new Error('No array buffer provided.');
         }
 
         const view = new DataView(buffer);
         // Format: X1,Y1,Z1,R1,G1,B1,A1,[...],XN,YN,ZN,RN,GN,BN,AN
-        const numPoints = Math.floor(buffer.byteLength / 16);
-
-        const positions = new Float32Array(3 * numPoints);
-        const colors = new Uint8Array(4 * numPoints);
+        let pointByteSize = 0;
+        for (const potreeName of pointAttributes) {
+            const attr = POINT_ATTTRIBUTES[potreeName];
+            pointByteSize += attr.numElements * (attr.numByte || attr.arrayType.BYTES_PER_ELEMENT);
+        }
+        const numPoints = Math.floor(buffer.byteLength / pointByteSize);
 
-        const tightbbox = new THREE.Box3();
-        tightbbox.min.set(Infinity, Infinity, Infinity);
-        tightbbox.max.set(-Infinity, -Infinity, -Infinity);
-        const tmp = new THREE.Vector3();
+        const attrs = [];
+        // get the variable attributes
+        for (const potreeName of pointAttributes) {
+            const attr = POINT_ATTTRIBUTES[potreeName];
+            attrs.push({
+                potreeName,
+                numElements: attr.numElements,
+                attributeName: attr.attributeName,
+                normalized: attr.normalized,
+                array: new attr.arrayType(attr.numElements * numPoints),
+                numByte: attr.numByte,
+            });
+        }
 
         let offset = 0;
-        for (let i = 0; i < numPoints; i++) {
-            positions[3 * i] = view.getUint32(offset + 0, true);
-            positions[3 * i + 1] = view.getUint32(offset + 4, true);
-            positions[3 * i + 2] = view.getUint32(offset + 8, true);
+        for (let pntIdx = 0; pntIdx < numPoints; pntIdx++) {
+            for (const attr of attrs) {
+                const fnName = `getUint${(attr.numByte || attr.array.BYTES_PER_ELEMENT) * 8}`;
```

Comment:
> this could move out of the pntIdx loop, by setting it as a function of attr in its definition loop.

What do we gain by that?

Reply:
> Maybe a comment on why getUint* works even on all types (float, Int...) (It looks strange to me)

it doesn't work on all types, it's just that potree converter stores everything as int, then putting an int into a Float32Array will convert it into a float.

Reply:
> What do we gain by that?

`numPoints` times less evaluations of ```view[`getUint${(attr.numByte || attr.array.BYTES_PER_ELEMENT) * 8}`]``` (and cleaner code IMHO :))

Reply:
> by setting it as a function of attr in its definition loop.

you mean as a property of attr then?

Reply:
> it doesn't work on all types, it's just that potree converter stores everything as int, then putting an int into a Float32Array will convert it into a float.

Indeed, by looking into it, potree is relying on a [custom view object](https://github.com/potree/potree/blob/develop/src/workers/BinaryDecoderWorker.js) rather than a DataView. I am not sure the 2 approaches are strictly equivalent (It is not clear to me who is manipulating bytes and who is doing type casting), but if your approach works, then let us go with it !

Reply:
> you mean as a property of attr then?

yes